test/core/event_engine/posix/posix_endpoint_test.cc

```diff
@@ -0,0 +1,410 @@
+#include "src/core/lib/event_engine/posix_engine/tcp_socket_utils.h"
+#include "src/core/lib/gprpp/dual_ref_counted.h"
+#include "src/core/lib/gprpp/global_config.h"
+#include "src/core/lib/iomgr/port.h"
+#include "test/core/event_engine/test_suite/event_engine_test_utils.h"
+#include "test/core/event_engine/test_suite/oracle_event_engine_posix.h"
+#include "test/core/util/port.h"
+
+GPR_GLOBAL_CONFIG_DECLARE_STRING(grpc_poll_strategy);
+
+namespace grpc_event_engine {
+namespace posix_engine {
+
+using ::grpc_event_engine::experimental::ChannelArgsEndpointConfig;
+using ::grpc_event_engine::experimental::EventEngine;
+using ::grpc_event_engine::experimental::Poller;
+using ::grpc_event_engine::experimental::PosixEventEngine;
+using ::grpc_event_engine::experimental::PosixOracleEventEngine;
+using ::grpc_event_engine::experimental::Promise;
+using ::grpc_event_engine::experimental::URIToResolvedAddress;
+using Endpoint = ::grpc_event_engine::experimental::EventEngine::Endpoint;
+using Listener = ::grpc_event_engine::experimental::EventEngine::Listener;
+using namespace std::chrono_literals;
+
+namespace {
+
+constexpr int kMinMessageSize = 1024;
+constexpr int kMaxMessageSize = 8192;
+constexpr int kNumConnections = 10;
+constexpr int kNumExchangedMessages = 100;
+std::atomic<int> g_num_active_connections{0};
+
+// Returns a random message with bounded length.
+std::string GetNextSendMessage() {
+  static const char alphanum[] =
+      "0123456789"
+      "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+      "abcdefghijklmnopqrstuvwxyz";
+  static std::random_device rd;
+  static std::seed_seq seed{rd()};
+  static std::mt19937 gen(seed);
+  static std::uniform_real_distribution<> dis(kMinMessageSize, kMaxMessageSize);
+  static grpc_core::Mutex g_mu;
+  std::string tmp_s;
+  int len;
+  {
+    grpc_core::MutexLock lock(&g_mu);
+    len = dis(gen);
+  }
+  tmp_s.reserve(len);
+  for (int i = 0; i < len; ++i) {
+    tmp_s += alphanum[rand() % (sizeof(alphanum) - 1)];
+  }
+  return tmp_s;
+}
+
+std::shared_ptr<EventEngine> GetPosixEE() {
+  static std::shared_ptr<EventEngine> posix_ee =
+      std::make_shared<PosixEventEngine>();
+  return posix_ee;
+}
+
+EventEngine* GetOracleEE() {
+  static EventEngine* oracle_ee = new PosixOracleEventEngine();
+  return oracle_ee;
+}
+
+class TestScheduler : public Scheduler {
+ public:
+  explicit TestScheduler(EventEngine* engine) : engine_(engine) {}
+  void Run(EventEngine::Closure* closure) override { engine_->Run(closure); }
+
+  void Run(absl::AnyInvocable<void()> cb) override {
+    engine_->Run(std::move(cb));
+  }
+
+ private:
+  EventEngine* engine_;
+};
+
+std::list<std::tuple<std::unique_ptr<EventEngine::Endpoint>,
+                     std::unique_ptr<EventEngine::Endpoint>>>
+CreateConnectedEndpoints(PosixEventPoller* poller, bool is_zero_copy_enabled,
+                         int num_connections) {
+  EXPECT_NE(GetOracleEE(), nullptr);
+  EXPECT_NE(GetPosixEE(), nullptr);
+  EXPECT_NE(poller, nullptr);
```

Comment:
If the argument type was `const PosixEvenPoller& poller`, the compiler would enforce this constraint (modulo some funny business).

Reply:
I think its better left as a pointer because its also captured inside a closure in line 133.

Reply:
Are you concerned about the object's lifetime, or the capture itself? You can capture it by reference for the same cost, I believe ... `[&poller] { poller.Kick(); }`. Regarding lifetimes, I don't think pointer or reference matters here since ownership is not transferred ... the caller must ensure it's alive until the callback is done.

Reply:
Fixed it now in the latest commit to use a reference. I couldn't use a const reference though because the function calls poller.Kick() which cannot be marked const.